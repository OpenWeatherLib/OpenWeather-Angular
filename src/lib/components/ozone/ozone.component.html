<mat-progress-bar *ngIf="(isLoading$ | async)" mode="indeterminate"></mat-progress-bar>
<mat-card class="ga-card" *ngIf="(ozone$ | async) && !(isLoading$ | async)">
  <mat-card-header>
    <img mat-card-avatar class="ga-card-header-image" src="/assets/ozone.png" alt="Ozone pic">
    <mat-card-title>Ozone</mat-card-title>
    <mat-card-subtitle>{{(ozone$ | async)?.data}}</mat-card-subtitle>
  </mat-card-header>
  <mat-card-content>
    <p>Coordinates: Lat: {{(ozone$ | async)?.coordinates?.latitude}}, Lon: {{(ozone$ | async)?.coordinates?.longitude}}</p>
  </mat-card-content>
</mat-card>
