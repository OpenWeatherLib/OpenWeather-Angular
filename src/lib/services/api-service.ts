import { DownloadResult, DownloadType } from "../enums";
import { City } from "../models";

export class ApiService {
  private currentWeatherUrl: string = "http://api.openweathermap.org/data/2.5/weather?q=%s&units=metric&APPID=%s"
  private forecastWeatherUrl: string = "http://api.openweathermap.org/data/2.5/forecast?q=%s&units=metric&APPID=%s"
  private uvIndexUrl: string = "http://api.openweathermap.org/data/2.5/uvi?lat=%.2f&lon=%.2f&APPID=%s"

  constructor(city: City, apiKey: string) { }

  currentWeather(): DownloadResult {
    //TODO
    return DownloadResult.Null;
  }

  forecastWeather(): DownloadResult {
    //TODO
    return DownloadResult.Null;
  }

  uvIndex(): DownloadResult {
    //TODO
    return DownloadResult.Null;
  }

  private doRestCall(downloadType: DownloadType, url: string): DownloadResult {
    return DownloadResult.Null
  }
}
