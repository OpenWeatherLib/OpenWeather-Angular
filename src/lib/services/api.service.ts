import { Injectable } from "@angular/core";

import { environment } from "../../environments/environment";

import { DownloadResult, DownloadType } from "../enums";
import { City } from "../models";

@Injectable()
export class ApiService {
  private currentWeatherUrl: string = "http://api.openweathermap.org/data/2.5/weather?q=%s&units=metric&APPID=%s"
  private forecastWeatherUrl: string = "http://api.openweathermap.org/data/2.5/forecast?q=%s&units=metric&APPID=%s"
  private uvIndexUrl: string = "http://api.openweathermap.org/data/2.5/uvi?lat=%.2f&lon=%.2f&APPID=%s"

  private apiKey = environment.apiKey;

  constructor(city: City) { }

  currentWeather(): DownloadResult {
    // TODO convert url and add lat, long, and apikey
    return this.doRestCall(DownloadType.Current, this.currentWeatherUrl);
  }

  forecastWeather(): DownloadResult {
    // TODO convert url and add lat, long, and apikey
    return this.doRestCall(DownloadType.Forecast, this.forecastWeatherUrl);
  }

  uvIndex(): DownloadResult {
    // TODO convert url and add lat, long, and apikey
    return this.doRestCall(DownloadType.UvIndex, this.uvIndexUrl);
  }

  private doRestCall(downloadType: DownloadType, url: string): DownloadResult {
    return DownloadResult.Null
  }
}
