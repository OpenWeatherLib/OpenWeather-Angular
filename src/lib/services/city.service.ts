import { Injectable } from "@angular/core";
import { Observable, of } from "rxjs";
import { map } from "rxjs/operators";

import { environment } from "@environments/environment";
import { validate, required } from "@lib/decorator";
import { ValidationRequiredType } from "@lib/enums";
import { any, format } from "@lib/helper";
import { City, City2 } from "@lib/models";
import { ApiService } from "@lib/services/api.service";

@Injectable()
export class CityService {

    constructor(private readonly apiService: ApiService) { }

    @validate(of(undefined))
    loadCityData(@required(ValidationRequiredType.String) cityName: string): Observable<City> {
        return this.apiService.get<any>(format(environment.urls.apis.geoCodeForCity, cityName), true)
            .pipe(map((response: any) => {
                if (!!response && !!response.status && response.status === "OK" && any(response.results)) {
                    const city2: City2 = response.results[0];

                    return {
                        id: -1,
                        name: city2.address_components[0].short_name,
                        country: city2.address_components[1].short_name,
                        population: -1,
                        coord: { lat: city2.geometry.location.lat, lon: city2.geometry.location.lng }
                    };
                }

                return undefined;
            }));
    }
}
