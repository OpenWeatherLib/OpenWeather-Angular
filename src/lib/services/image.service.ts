import { Injectable } from "@angular/core";
import { BehaviorSubject, Observable, of } from "rxjs";
import { catchError, map, take } from "rxjs/operators";

import "@lib/extensions/string.extensions";

import { environment } from "../../environments/environment";

import { required, validate } from "@lib/decorator";
import { UnsplashImageOrientation, ValidationRequiredType } from "@lib/enums";
import { UnsplashImageResponse } from "@lib/models";
import { ApiService } from "@lib/services/api.service";

@Injectable()
export class ImageService {

    private imageApiUrl: string = "https://api.unsplash.com/search/photos?client_id={0}&orientation={1}&query={2}";

    private accessKey: string = environment.unsplashAccessKey;

    private cityPictureUrl$ = new BehaviorSubject<string>(null);

    constructor(private readonly apiService: ApiService) { }

    cityPictureUrl(): Observable<string> {
        return this.cityPictureUrl$;
    }

    @validate(of(null))
    receiveImagePictureUrl(@required(ValidationRequiredType.String) cityName: string,
        orientation: UnsplashImageOrientation = UnsplashImageOrientation.Squarish): void {

        if (!this.accessKey) {
            return;
        }

        this.apiService.get<UnsplashImageResponse>(String().format(this.imageApiUrl, this.accessKey, orientation, cityName))
            .pipe(
                take(1),
                map(response => {
                    if (response && response.total && response.results) {
                        return response.results[0].urls.small;
                    }
                    return null;
                }),
                catchError(error => {
                    console.error(error.toString());
                    return null;
                }))
            .subscribe(response => {
                if (response) {
                    this.cityPictureUrl$.next(response);
                }
            });
    }
}
