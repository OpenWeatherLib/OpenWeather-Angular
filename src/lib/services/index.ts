export * from "./api.service";
export * from "./image.service";
export * from "./open-weather.service";
