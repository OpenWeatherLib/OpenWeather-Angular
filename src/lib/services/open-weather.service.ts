import { Injectable } from "@angular/core";
import { BehaviorSubject, Observable } from "rxjs";
import { take } from "rxjs/operators";

import "@lib/extensions/string.extensions";

import { environment } from "../../environments/environment";

import { ApiCallState } from "@lib/enums";
import { City, City2, Coordinates, UvIndex, WeatherCurrent, WeatherForecast } from "@lib/models";
import { ApiService } from "@lib/services/api.service";

@Injectable()
export class OpenWeatherService {

    private city$ = new BehaviorSubject<City>(null);
    private currentWeather$ = new BehaviorSubject<WeatherCurrent>(null);
    private forecastWeather$ = new BehaviorSubject<WeatherForecast>(null);
    private uvIndex$ = new BehaviorSubject<UvIndex>(null);

    private apiKey: string = environment.apiKey;

    constructor(private readonly apiService: ApiService) { }

    city(): Observable<City> {
        return this.city$;
    }

    setCityData(name: string, coord?: Coordinates): void {
        const city = this.city$.value;
        city.name = name;
        if (coord) {
            city.coord = coord;
        }
        this.city$.next(city);
    }

    currentWeather(): Observable<WeatherCurrent> {
        return this.currentWeather$;
    }

    forecastWeather(): Observable<WeatherForecast> {
        return this.forecastWeather$;
    }

    uvIndex(): Observable<UvIndex> {
        return this.uvIndex$;
    }

    loadCityData(cityName: string): void {
        this.apiService.geoCodeForCity(cityName)
            .pipe(take(1))
            .subscribe(response => {
                if (response && response.status && response.status === "OK" && response.results && response.results.length > 0) {
                    const city2: City2 = response.results[0];
                    this.setCityData(city2.address_components[0].long_name, { lat: city2.geometry.location.lat, lon: city2.geometry.location.lng });
                }
            });
    }

    loadCurrentWeather(): ApiCallState {
        if (!this.city$.value || !this.city$.value.isNameSet()) {
            return ApiCallState.CityNotSet;
        }

        this.apiService.currentWeather(this.apiKey, this.city$.value)
            .pipe(take(1))
            .subscribe(response => {
                if (response) {
                    this.currentWeather$.next(response);
                }
            });

        return ApiCallState.Calling;
    }

    loadForecastWeather(): ApiCallState {
        if (!this.city$.value || !this.city$.value.isCoordSet()) {
            return ApiCallState.CoordNotSet;
        }

        this.apiService.forecastWeather(this.apiKey, this.city$.value)
            .pipe(take(1))
            .subscribe(response => {
                if (response) {
                    this.forecastWeather$.next(response);
                }
            });

        return ApiCallState.Calling;
    }

    loadUvIndex(): ApiCallState {
        if (!this.city$.value || !this.city$.value.isCoordSet()) {
            return ApiCallState.CoordNotSet;
        }

        this.apiService.uvIndex(this.apiKey, this.city$.value)
            .pipe(take(1))
            .subscribe(response => {
                if (response) {
                    this.uvIndex$.next(response);
                }
            });

        return ApiCallState.Calling;
    }
}
