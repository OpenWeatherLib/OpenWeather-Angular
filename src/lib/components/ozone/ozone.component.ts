import { Component, OnInit } from "@angular/core";

import { BaseComponent } from "@lib/components/base-component/base.component";
import { Ozone } from "@lib/models";
import { OpenWeatherService } from "@lib/services";

@Component({
  selector: "ga-ozone",
  templateUrl: "./ozone.component.html",
  styleUrls: ["./ozone.component.scss"]
})
export class OzoneComponent extends BaseComponent implements OnInit {

  ozone: Ozone = null;

  constructor(private readonly openWeatherService: OpenWeatherService) {
    super();
  }

  ngOnInit() {
    this.registerSubscription(this.openWeatherService.ozone().subscribe(ozone => {
      if (ozone) {
        this.ozone = ozone;
      }
    }));
  }
}
