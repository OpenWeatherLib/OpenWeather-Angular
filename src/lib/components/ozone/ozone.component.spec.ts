import { TestBed, async } from "@angular/core/testing";
import { of } from "rxjs";

import MockServices from "@lib/mock/services.mock";
// import MockValues from "@lib/mock/values.mock";

import { OpenWeatherService } from "@lib/services";

import { OzoneComponent } from "./ozone.component";

describe("OzoneComponent", () => {
  let classToTest: OzoneComponent;

  const openWeatherServiceMock = MockServices.substitute(OpenWeatherService);

  const serviceMockList: any[] = [
    openWeatherServiceMock
  ];

  beforeEach(() => {
    openWeatherServiceMock.ozone.and.returnValue(of(null));

    TestBed.configureTestingModule({
      declarations: [
        OzoneComponent
      ],
      providers: [
        { provide: OpenWeatherService, useValue: openWeatherServiceMock }
      ]
    }).overrideTemplate(OzoneComponent, "<div></div>").compileComponents();

    const fixture = TestBed.createComponent(OzoneComponent);
    classToTest = fixture.debugElement.componentInstance;
  });

  afterEach(() => {
    serviceMockList.forEach(serviceMock => {
      for (const propertyName in serviceMock) {
        if (serviceMock.hasOwnProperty(propertyName)) {
          serviceMock[propertyName].calls.reset();
          serviceMock[propertyName].and.stub();
        }
      }
    });
  });

  it("should create the app", async(() => {
    // Assert
    expect(classToTest).toBeTruthy();
  }));
});
