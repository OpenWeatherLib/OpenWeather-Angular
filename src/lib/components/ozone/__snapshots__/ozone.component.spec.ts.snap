// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OzoneComponent ngOnInit ozone$ 1`] = `
Object {
  "coordinates": Object {
    "latitude": 0,
    "longitude": 10,
  },
  "data": 259.3334655761719,
  "dateTime": Object {},
}
`;
